### DISCUSSION TITLE
Polyphonic playback from multiple sources - is it possible with the Audio lib?

### DISCUSSION THREAD

**Discussion by Protonerd** [schreibfaul1/ESP32-audioI2S#462] [Q&A]
First off thanks a lot for the tremendous effort you and all supporters put into this project! Special thanks for the quick feedback in case of issues, thumbs up!
An idea I have for an ESP32 based project involves playing multiple sound files mixed together at the same time. Ideally these sound files should be able to come from different sources, like streaming a web radio and mixing in a wav file stored on an SD-card. Is it possible with this library?

**Comment by schreibfaul1** [comment]
Thank you for your contribution. I don't intend to open a second signal source any time soon. But I could imagine using the 2nd I2S channel as an input for a microphone or opening a second file on the SD card (SD_MMC allows this) and adding up the signals.

**Comment by Protonerd** [comment]
Thanks Wolle for the the reply. We in fact already made a wavplayer capable of polyphonic feedback for the nRF52 chipset and now try to port the code to the ESP32. On the nRF the driver was quite simple, so a double buffering scheme had to be implemented and alternatingly fed to the I2S, once one of the buffer content was consumed. Reading in multiple files from the SD can be done and then all sources added together (weighted with individually selectable volume). At the end the mixed buffer needs to be fed to the I2S. On the ESP32 the driver seems more capable, but also not easy to understand (for me at least). I've been spending a lot of time lately to analyse your excellent library, but I'm still not sure I understand how to pass the samples to the I2S. I made a double buffer with 512 samples each in the config (dma_buf_count and dma_buf_len), I guess this natively replaces the double buffering. However as to how to pass it to I2S I'm in doubt:
- I assumed that the function i2s_write can be fed with the whole data from red from the SD-card (already in i2s format). But it does not work. Looking at your code it seems to me that in playSample you pass only one sample to the i2s_write function.
- I also do not find how to determine when the i2s driver finished playing all passed samples in the internal buffer and ready to receive new ones.
Would be very greateful if you could help shed some lights on these topics...

**Reply by schreibfaul1** [reply]
There is a function 'playChunk()' that passes the audio data present in a buffer to I2S. The function differentiates between mono/stereo or 8/16bit. If there is a second buffer to read, each sample from the two sources can be added together

**Reply by Protonerd** [reply]
Thanks Wolle, going further in the code the playChunk() calls playSample, there it is not yet clear to me how the address of the sample passed via the i2s_write function is calculated. It is the variable s32. Does it refer to an address in the inBuff?
Another question is, if the i2s buffer is selected to be 8k (5*1600 bytes in case of wav files), why are the samples passed one by one instead of reading/streaming in the whole 8k and passing it in one go to i2s driver?

**Comment by probonopd** [comment]
I would also be interested in this. My use case would be mixing the station ID (e.g., from TTS or a SPIFFS file) into a running web radio stream when switching channels.
